File: src/plot_astrometric_residuals_xy.c
// plot_astrometric_residuals_xy -- PGPLOT diagnostic for catalog-matched
// star distribution across an image.
//
// Input:  a FITS file (e.g., wcs_<image>.fts) or its accompanying
//         .cat.astrometric_residuals file.
// Output: <basename>_astrometric_residuals.png in the current directory.
//
// The .cat.astrometric_residuals file is produced by solve_plate_with_UCAC5
// (write_astrometric_residuals_vector_field). Columns 8 and 9 are x_pix,
// y_pix of each catalog-matched star. This tool plots those coordinates
// with the plot's aspect ratio matched to the source image's NAXIS1:NAXIS2,
// so the distribution is visually faithful to the chip layout.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "fitsio.h"
#include "cpgplot.h"

#include "setenv_local_pgplot.h"

#define MAX_POINTS 1000000
#define PATH_MAX_LEN 4096

// Strip a trailing .fts / .fits / .fit extension (case-insensitive) and
// return the index where the extension started; returns the original length
// if no recognized extension is present.
static size_t length_without_fits_ext( const char *path ) {
 size_t len;
 size_t i;
 const char *suffix;
 size_t suffix_len;
 const char *known_suffixes[] = { ".fits", ".fts", ".fit", NULL };
 len= strlen( path );
 for ( i= 0; known_suffixes[i] != NULL; i++ ) {
  suffix= known_suffixes[i];
  suffix_len= strlen( suffix );
  if ( len >= suffix_len ) {
   size_t j;
   int matches= 1;
   for ( j= 0; j < suffix_len; j++ ) {
    char a;
    char b;
    a= path[len - suffix_len + j];
    b= suffix[j];
    if ( a >= 'A' && a <= 'Z' )
     a= (char)( a - 'A' + 'a' );
    if ( a != b ) {
     matches= 0;
     break;
    }
   }
   if ( matches )
    return len - suffix_len;
  }
 }
 return len;
}

// Strip the trailing ".cat.astrometric_residuals" if present; returns the
// original length otherwise.
static size_t length_without_residuals_ext( const char *path ) {
 const char *suffix = ".cat.astrometric_residuals";
 size_t len;
 size_t suffix_len;
 len= strlen( path );
 suffix_len= strlen( suffix );
 if ( len >= suffix_len && strcmp( path + len - suffix_len, suffix ) == 0 )
  return len - suffix_len;
 return len;
}

// Read NAXIS1 / NAXIS2 from the FITS file at fits_path. Returns 0 on
// success, non-zero on failure. Uses fits_open_image so compressed files
// (.fz) are handled transparently.
static int read_fits_dimensions( const char *fits_path, long *nx, long *ny ) {
 fitsfile *fptr= NULL;
 int status= 0;
 int naxis= 0;
 long naxes[9];
 fits_open_image( &fptr, (char *)fits_path, READONLY, &status );
 if ( status != 0 ) {
  fits_report_error( stderr, status );
  fits_clear_errmsg();
  return 1;
 }
 fits_get_img_dim( fptr, &naxis, &status );
 if ( status != 0 ) {
  fits_report_error( stderr, status );
  fits_clear_errmsg();
  fits_close_file( fptr, &status );
  return 2;
 }
 if ( naxis < 2 ) {
  fprintf( stderr, "plot_astrometric_residuals_xy: %s has NAXIS=%d, expected >=2\n",
           fits_path, naxis );
  fits_close_file( fptr, &status );
  return 3;
 }
 fits_get_img_size( fptr, 2, naxes, &status );
 if ( status != 0 ) {
  fits_report_error( stderr, status );
  fits_clear_errmsg();
  fits_close_file( fptr, &status );
  return 4;
 }
 fits_close_file( fptr, &status );
 *nx= naxes[0];
 *ny= naxes[1];
 return 0;
}

// Read column 8 (x_pix) and column 9 (y_pix) from each line. Returns the
// number of points successfully read or -1 on file-open failure.
static int read_xy_from_residuals_file( const char *path, float *x, float *y, int max_points ) {
 FILE *f;
 char line[1024];
 int n;
 double c[9];
 f= fopen( path, "r" );
 if ( f == NULL ) {
  return -1;
 }
 n= 0;
 while ( n < max_points && fgets( line, (int)sizeof( line ), f ) != NULL ) {
  int parsed;
  parsed= sscanf( line, "%lf %lf %lf %lf %lf %lf %lf %lf %lf",
                  &c[0], &c[1], &c[2], &c[3], &c[4], &c[5], &c[6], &c[7], &c[8] );
  if ( parsed == 9 ) {
   x[n]= (float)c[7];
   y[n]= (float)c[8];
   n++;
  }
 }
 fclose( f );
 return n;
}

int main( int argc, char **argv ) {
 char fits_path[PATH_MAX_LEN];
 char residuals_path[PATH_MAX_LEN];
 char png_path[PATH_MAX_LEN];
 char png_device_spec[PATH_MAX_LEN];
 char title[PATH_MAX_LEN];
 char fits_basename[PATH_MAX_LEN];
 long nx;
 long ny;
 int have_fits;
 size_t input_len;
 const char *input;
 float *x;
 float *y;
 int n_points;
 int dim_status;
 size_t base_len;
 const char *bn;
 size_t i;
 double aspect_ratio;

 if ( argc != 2 ) {
  fprintf( stderr, "Usage: %s <fits-file | .cat.astrometric_residuals>\n", argv[0] );
  fprintf( stderr, "Plots the (x_pix, y_pix) distribution of catalog-matched\n" );
  fprintf( stderr, "stars from the .cat.astrometric_residuals file accompanying\n" );
  fprintf( stderr, "the given FITS image.\n" );
  return 1;
 }
 input= argv[1];
 input_len= strlen( input );
 if ( input_len >= PATH_MAX_LEN - 32 ) {
  fprintf( stderr, "%s: input path too long\n", argv[0] );
  return 1;
 }

 // Determine which side of the (FITS, residuals) pair we were given.
 if ( input_len >= strlen( ".cat.astrometric_residuals" ) &&
      strcmp( input + input_len - strlen( ".cat.astrometric_residuals" ),
              ".cat.astrometric_residuals" ) == 0 ) {
  // Input is the residuals file; derive the FITS path by stripping the
  // ".cat.astrometric_residuals" suffix.
  size_t fits_len;
  fits_len= length_without_residuals_ext( input );
  memcpy( fits_path, input, fits_len );
  fits_path[fits_len]= '\0';
  strncpy( residuals_path, input, PATH_MAX_LEN - 1 );
  residuals_path[PATH_MAX_LEN - 1]= '\0';
 } else {
  // Input is the FITS file; the residuals file sits beside it with the
  // ".cat.astrometric_residuals" suffix.
  strncpy( fits_path, input, PATH_MAX_LEN - 1 );
  fits_path[PATH_MAX_LEN - 1]= '\0';
  snprintf( residuals_path, PATH_MAX_LEN, "%s.cat.astrometric_residuals", fits_path );
 }

 // Try to read NAXIS1 / NAXIS2. If the FITS file is not available, we will
 // fall back to bounds from the residuals data later.
 have_fits= ( 0 == read_fits_dimensions( fits_path, &nx, &ny ) );
 if ( !have_fits ) {
  fprintf( stderr,
           "plot_astrometric_residuals_xy: could not read dimensions from %s, "
           "falling back to data extent\n",
           fits_path );
 }

 // Allocate point buffers and read the residuals file.
 x= (float *)malloc( (size_t)MAX_POINTS * sizeof( float ) );
 y= (float *)malloc( (size_t)MAX_POINTS * sizeof( float ) );
 if ( x == NULL || y == NULL ) {
  fprintf( stderr, "plot_astrometric_residuals_xy: malloc failed\n" );
  free( x );
  free( y );
  return 2;
 }
 n_points= read_xy_from_residuals_file( residuals_path, x, y, MAX_POINTS );
 if ( n_points < 0 ) {
  // Friendly to the calling pipeline: not finding the residuals file is not
  // a hard error since the plate solve may have failed for other reasons.
  fprintf( stderr,
           "plot_astrometric_residuals_xy: cannot open %s -- skipping plot\n",
           residuals_path );
  free( x );
  free( y );
  return 0;
 }

 // If we have no FITS dimensions, derive a bounding box from the data.
 if ( !have_fits ) {
  if ( n_points <= 0 ) {
   fprintf( stderr,
            "plot_astrometric_residuals_xy: %s is empty and no FITS dimensions; "
            "nothing to plot\n",
            residuals_path );
   free( x );
   free( y );
   return 0;
  } else {
   float xmin;
   float xmax;
   float ymin;
   float ymax;
   int j;
   xmin= xmax= x[0];
   ymin= ymax= y[0];
   for ( j= 1; j < n_points; j++ ) {
    if ( x[j] < xmin )
     xmin= x[j];
    if ( x[j] > xmax )
     xmax= x[j];
    if ( y[j] < ymin )
     ymin= y[j];
    if ( y[j] > ymax )
     ymax= y[j];
   }
   nx= (long)( xmax - xmin + 1.0f );
   ny= (long)( ymax - ymin + 1.0f );
   if ( nx < 1 )
    nx= 1;
   if ( ny < 1 )
    ny= 1;
  }
 }

 // Build the output PNG filename. Strip a trailing .fts/.fits/.fit from
 // the FITS basename so the result reads cleanly.
 bn= strrchr( fits_path, '/' );
 if ( bn == NULL )
  bn= fits_path;
 else
  bn= bn + 1;
 strncpy( fits_basename, bn, PATH_MAX_LEN - 1 );
 fits_basename[PATH_MAX_LEN - 1]= '\0';
 base_len= length_without_fits_ext( fits_basename );
 fits_basename[base_len]= '\0';
 // Replace characters PGPLOT's /PNG device interprets specially in the
 // filename portion -- the device spec ends at '/', so any '/' or '+' in
 // the cwd-relative output name would confuse it. Our convention is to
 // write to cwd so the name should be safe, but be defensive.
 for ( i= 0; i < base_len; i++ ) {
  if ( fits_basename[i] == '/' )
   fits_basename[i]= '_';
 }
 snprintf( png_path, PATH_MAX_LEN, "%s_astrometric_residuals.png", fits_basename );

 // PGPLOT device specification: <filename>/PNG.
 snprintf( png_device_spec, PATH_MAX_LEN, "%s/PNG", png_path );

 // Title shows the FITS basename and the matched-star count.
 snprintf( title, PATH_MAX_LEN, "%s  [%d matched]", fits_basename, n_points );

 // Aspect ratio for cpgpap: PGPLOT interprets the second argument as
 // height/width. We want a plot whose drawing area is proportional to the
 // image (NAXIS2 / NAXIS1).
 aspect_ratio= (double)ny / (double)nx;
 if ( aspect_ratio <= 0.0 )
  aspect_ratio= 1.0;

 setenv_localpgplot( argv[0] );
 if ( 1 == cpgbeg( 0, png_device_spec, 1, 1 ) ) {
  // White background, black foreground -- match fit_mag_calib's convention
  // so the resulting PNG is readable on the white-background HTML log.
  cpgscr( 0, 1.0f, 1.0f, 1.0f );
  cpgscr( 1, 0.0f, 0.0f, 0.0f );
  // Default page width with the requested aspect ratio.
  cpgpap( 8.0, (float)aspect_ratio );
  cpgsvp( 0.10f, 0.95f, 0.10f, 0.92f );
  cpgswin( 0.0f, (float)nx, 0.0f, (float)ny );
  cpgscf( 1 );
  cpgbox( "BCNST", 0.0, 0, "BCNST", 0.0, 0 );
  cpglab( "X (pixels)", "Y (pixels)", title );
  if ( n_points > 0 ) {
   // Filled circle marker in red (color index 2), slightly smaller than
   // default so dense fields remain readable.
   cpgsch( 0.6f );
   cpgsci( 2 );
   cpgpt( n_points, x, y, 17 );
  }
  cpgend();
  fprintf( stderr,
           "plot_astrometric_residuals_xy: wrote %s (%d matched stars from %s, image %ldx%ld)\n",
           png_path, n_points, residuals_path, nx, ny );
 } else {
  fprintf( stderr,
           "plot_astrometric_residuals_xy: PGPLOT /PNG device not available -- "
           "no plot produced\n" );
 }

 free( x );
 free( y );
 return 0;
}

File: lib/plot_astrometric_residuals_xy.py
#!/usr/bin/env python3
#
# plot_astrometric_residuals_xy.py -- informative diagnostic of the astrometric
# residual vector field across an image.
#
# This is the preferred (richer) replacement for the PGPLOT C tool
# lib/plot_astrometric_residuals_xy, which only plots the (x_pix, y_pix)
# distribution of catalog-matched stars. Here we additionally draw:
#   - the residual VECTOR FIELD (each matched star's measured-minus-catalog
#     offset, exaggerated) coloured by the offset from the catalog position, and
#   - an offset-from-catalog map,
# so a glance tells you whether the WCS is good (small, random residuals) or
# bad (large, spatially-structured residuals -- e.g. a distortion the solution
# failed to model, which throws off the RA/Dec of every detected source and
# makes real stars look like unidentified transients).
#
# Note: throughout this tool "offset" / "residual" means the angular distance
# between a star's measured (WCS) position and its catalog position, in
# arcseconds -- NOT a stellar brightness magnitude.
#
# Input/Output mirror the C tool exactly, so the calling pipeline can use
# either interchangeably:
#   Input : <image>.fits           (residuals read from <image>.fits.wcscat.astrometric_residuals)
#        or <image>.fits.wcscat.astrometric_residuals
#   Output: <image>_astrometric_residuals.png   in the current directory.
#
# The residuals file is written by util/solve_plate_with_UCAC5
# (write_astrometric_residuals_vector_field). Whitespace-separated columns are:
#   1 RA(deg) 2 Dec(deg) 3 dRA(deg) 4 dDec(deg) 5 |resid|(arcsec)
#   6 dRA*cos(Dec)(arcsec) 7 dDec(arcsec) 8 x_pix 9 y_pix
#
# Dependencies: numpy + matplotlib only. If they are missing the script exits
# non-zero with a one-line message so the caller can fall back to the C tool.
# No astropy needed -- image dimensions are read from the FITS header by a
# minimal parser, falling back to the data extent if that fails.

import os
import sys

# Column indices (0-based) in the .wcscat.astrometric_residuals file.
COL_OFFSET = 4      # |resid| = offset from catalog position, arcsec
COL_DX = 5          # dRA*cos(Dec) in arcsec (East-West residual)
COL_DY = 6          # dDec in arcsec (North-South residual)
COL_X = 7           # x_pix
COL_Y = 8           # y_pix
MIN_COLUMNS = 9


def strip_fits_ext(name):
    low = name.lower()
    for ext in (".fits", ".fts", ".fit"):
        if low.endswith(ext):
            return name[: len(name) - len(ext)]
    return name


def derive_paths(arg):
    """Return (residuals_path, fits_path) from either kind of input path."""
    suffix = ".wcscat.astrometric_residuals"
    if arg.endswith(suffix):
        residuals_path = arg
        fits_path = arg[: -len(suffix)]
    else:
        fits_path = arg
        residuals_path = arg + ".wcscat.astrometric_residuals"
    return residuals_path, fits_path


def read_fits_dimensions(fits_path):
    """Read NAXIS1/NAXIS2 from a FITS primary header without astropy.

    Returns (nx, ny) or (None, None) on any failure (caller falls back to the
    data extent). Only the uncompressed primary header is parsed, which is what
    the wcs_-prefixed images written by the pipeline carry.
    """
    try:
        nx = ny = None
        with open(fits_path, "rb") as f:
            while True:
                block = f.read(2880)
                if len(block) < 2880:
                    break
                done = False
                for off in range(0, 2880, 80):
                    card = block[off:off + 80].decode("ascii", "replace")
                    key = card[:8].strip()
                    if key == "END":
                        done = True
                        break
                    if key in ("NAXIS1", "NAXIS2") and card[8:10] == "= ":
                        try:
                            val = int(card[10:].split("/")[0].strip())
                        except ValueError:
                            continue
                        if key == "NAXIS1":
                            nx = val
                        else:
                            ny = val
                if done or (nx is not None and ny is not None):
                    break
        if nx and ny and nx > 0 and ny > 0:
            return nx, ny
    except (IOError, OSError):
        pass
    return None, None


def read_residuals(residuals_path):
    """Return numpy arrays x, y, dx, dy, offset. Raises IOError if unreadable."""
    import numpy as np

    xs, ys, dxs, dys, offsets = [], [], [], [], []
    with open(residuals_path, "r") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            parts = line.split()
            if len(parts) < MIN_COLUMNS:
                continue
            try:
                xs.append(float(parts[COL_X]))
                ys.append(float(parts[COL_Y]))
                dxs.append(float(parts[COL_DX]))
                dys.append(float(parts[COL_DY]))
                offsets.append(float(parts[COL_OFFSET]))
            except ValueError:
                continue
    return (np.asarray(xs), np.asarray(ys), np.asarray(dxs),
            np.asarray(dys), np.asarray(offsets))


def mad_sigma(values):
    """Robust sigma estimate: 1.4826 * MAD. Returns 0.0 for empty input."""
    import numpy as np
    if values.size == 0:
        return 0.0
    med = np.median(values)
    return float(1.4826 * np.median(np.abs(values - med)))


def main():
    if len(sys.argv) != 2:
        sys.stderr.write(
            "Usage: %s <fits-file | .wcscat.astrometric_residuals>\n" % sys.argv[0])
        return 1

    # Import the heavy dependencies inside a guard so a missing package is a
    # clean, quiet fallback signal rather than a traceback.
    try:
        import numpy as np
        import matplotlib
        matplotlib.use("Agg")
        import matplotlib.pyplot as plt
    except ImportError as exc:
        sys.stderr.write(
            "plot_astrometric_residuals_xy.py: numpy/matplotlib unavailable "
            "(%s) -- falling back to the C plotter\n" % exc)
        return 2

    residuals_path, fits_path = derive_paths(sys.argv[1])

    if not os.path.isfile(residuals_path):
        sys.stderr.write(
            "plot_astrometric_residuals_xy.py: residuals file not found: %s "
            "-- nothing to plot\n" % residuals_path)
        return 1

    try:
        x, y, dx, dy, offset = read_residuals(residuals_path)
    except (IOError, OSError) as exc:
        sys.stderr.write(
            "plot_astrometric_residuals_xy.py: cannot read %s (%s)\n"
            % (residuals_path, exc))
        return 1

    n_points = x.size

    nx, ny = read_fits_dimensions(fits_path)
    if nx is None:
        if n_points > 0:
            nx = float(np.max(x)) + 1.0
            ny = float(np.max(y)) + 1.0
        else:
            nx = ny = 1.0

    base = strip_fits_ext(os.path.basename(fits_path)).replace("/", "_")
    png_path = base + "_astrometric_residuals.png"

    # Summary statistics that make the WCS quality self-evident.
    if n_points > 0:
        med_resid = float(np.median(offset))
        max_resid = float(np.max(offset))
        sigma = np.hypot(mad_sigma(dx), mad_sigma(dy))
    else:
        med_resid = max_resid = sigma = 0.0

    if n_points == 0:
        verdict = "NO MATCHED STARS -- plate solution failed for this image"
    elif med_resid < 1.5:
        verdict = "residuals small and random -> good astrometric solution"
    elif med_resid < 5.0:
        verdict = "residuals elevated -> marginal solution"
    else:
        verdict = ("residuals LARGE/structured -> bad WCS; source RA/Dec are "
                   "unreliable (real stars may appear as unidentified transients)")

    # Colour scale capped at the 98th percentile so a few large outliers do
    # not wash out the structure of the bulk of the field.
    if n_points > 0:
        vmax = float(np.percentile(offset, 98))
        if vmax <= 0:
            vmax = max(max_resid, 1e-6)
    else:
        vmax = 1.0

    fig, axes = plt.subplots(1, 2, figsize=(15, 6))

    # ---- Panel A: binned residual vector field ------------------------------
    # Average the per-star residuals into a coarse grid so the vector field is
    # readable (hundreds of overlapping arrows are not). A smooth, organised
    # pattern that grows toward one corner is the signature of an unmodelled
    # distortion; small, random arrows mean a clean solution.
    axA = axes[0]
    if n_points > 0:
        ncx, ncy = 26, 17
        ex = np.linspace(0, nx, ncx + 1)
        ey = np.linspace(0, ny, ncy + 1)
        ix = np.clip(np.digitize(x, ex) - 1, 0, ncx - 1)
        iy = np.clip(np.digitize(y, ey) - 1, 0, ncy - 1)
        gx, gy, gdx, gdy, goffset = [], [], [], [], []
        for cy in range(ncy):
            for cx in range(ncx):
                sel = (ix == cx) & (iy == cy)
                if not np.any(sel):
                    continue
                gx.append(0.5 * (ex[cx] + ex[cx + 1]))
                gy.append(0.5 * (ey[cy] + ey[cy + 1]))
                gdx.append(float(np.mean(dx[sel])))
                gdy.append(float(np.mean(dy[sel])))
                goffset.append(float(np.mean(offset[sel])))
        gx = np.asarray(gx); gy = np.asarray(gy)
        gdx = np.asarray(gdx); gdy = np.asarray(gdy); goffset = np.asarray(goffset)
        # Scale so a typical (cell-mean) arrow spans roughly one grid cell.
        ref = float(np.median(np.hypot(gdx, gdy))) if gx.size else med_resid
        if ref <= 0:
            ref = max(med_resid, 1e-6)
        target_pix = 0.9 * (nx / ncx)
        scale = (ref / target_pix) if target_pix > 0 else 1.0
        q = axA.quiver(gx, gy, gdx, gdy, goffset, angles="xy", scale_units="xy",
                       scale=scale, cmap="inferno", width=0.004,
                       clim=(0, vmax))
        cb = fig.colorbar(q, ax=axA)
        cb.set_label("mean offset from catalog (arcsec)")
        key_len = max(round(ref, 1), 0.1)
        axA.quiverkey(q, 0.82, 1.02, key_len, '%g"' % key_len,
                      labelpos="E", coordinates="axes")
    else:
        axA.text(0.5, 0.5, "no matched stars", ha="center", va="center",
                 transform=axA.transAxes, color="red", fontsize=14)
    axA.set_xlim(0, nx)
    axA.set_ylim(0, ny)
    axA.set_aspect("equal")
    axA.set_xlabel("X (pixels)")
    axA.set_ylabel("Y (pixels)")
    axA.set_title("Astrometric residual vector field (grid-averaged)\n"
                  "arrow = measured position minus catalog position")

    # ---- Panel B: per-star offset-from-catalog map --------------------------
    axB = axes[1]
    if n_points > 0:
        sc = axB.scatter(x, y, c=offset, s=14, cmap="inferno",
                         vmin=0, vmax=vmax, edgecolors="none")
        cb2 = fig.colorbar(sc, ax=axB)
        cb2.set_label("offset from catalog position (arcsec)")
    else:
        axB.text(0.5, 0.5, "no matched stars", ha="center", va="center",
                 transform=axB.transAxes, color="red", fontsize=14)
    axB.set_xlim(0, nx)
    axB.set_ylim(0, ny)
    axB.set_aspect("equal")
    axB.set_xlabel("X (pixels)")
    axB.set_ylabel("Y (pixels)")
    axB.set_title("Catalog-matched stars coloured by offset from catalog position")

    fig.suptitle(
        "%s\n%d matched stars   median=%.2f\"   max=%.1f\"   sigma(MAD)=%.2f\"   "
        "image %gx%g\n%s"
        % (base, n_points, med_resid, max_resid, sigma, nx, ny, verdict),
        fontsize=11)
    fig.tight_layout(rect=(0, 0, 1, 0.93))

    try:
        fig.savefig(png_path, dpi=80)
    except (IOError, OSError) as exc:
        sys.stderr.write(
            "plot_astrometric_residuals_xy.py: failed to write %s (%s)\n"
            % (png_path, exc))
        plt.close(fig)
        return 1
    plt.close(fig)

    # Only claim success if a non-empty file actually landed on disk.
    if os.path.isfile(png_path) and os.path.getsize(png_path) > 0:
        sys.stderr.write(
            "plot_astrometric_residuals_xy.py: wrote %s (%d matched stars from %s, "
            "image %gx%g, median residual %.2f arcsec)\n"
            % (png_path, n_points, residuals_path, nx, ny, med_resid))
        return 0
    sys.stderr.write(
        "plot_astrometric_residuals_xy.py: no output file produced at %s\n"
        % png_path)
    return 1


if __name__ == "__main__":
    sys.exit(main())
